Modbus: Add support for Holding Registers to Binary Sensor

Nippey opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I have a 32-channel input ("N4DIM32") for modbus. This device exposes its internal states via holding registers.
(If required, I can attach the interface description provided by the seller.)
As binary_sensor currently only supports CALL_TYPE_COIL, CALL_TYPE_DISCRETE:

**ADDED: Support for CALL_TYPE_REGISTER_HOLDING.**

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24595
- Community link: https://community.home-assistant.io/t/modbus-binary-sensor-cannnot-read-from-holding-register/474938

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Nippey committed:
Update binary_sensor.py

Update handling of binary sensors to support reading from holding registers (command 0x03).

Nippey committed:
Update __init__.py

Update prser to accept holding registers as input type for binary sensors

Nippey committed:
Formatted using Black

janiversen commented:
Will take a look later.

janiversen reviewed (changes requested):
Test are missing, please amend the relevant test_file with tests that control the new functionality.

Nippey committed:
Update binary_sensor.py

Comply with pylint output

Nippey commented on homeassistant/components/modbus/binary_sensor.py:
```diff
@@ -21,6 +21,8 @@
     DataUpdateCoordinator,
 )
 
+from .const import CALL_TYPE_COIL, CALL_TYPE_DISCRETE, CALL_TYPE_REGISTER_HOLDING
```
Thanks for the heads-up! 
Removed `CALL_TYPE_REGISTER_HOLDING`, as not implicitly used.

janiversen commented on homeassistant/components/modbus/binary_sensor.py:
```diff
@@ -109,9 +111,16 @@ async def async_update(self, now: datetime | None = None) -> None:
+            if self._input_type in (CALL_TYPE_COIL, CALL_TYPE_DISCRETE):
+                self._attr_is_on = bool(result.bits[0] & 1)
+            else:
+                value = int(result.registers[0])
```
The holding register only holds 1 bit ??? I would have expected it to contain 16bits just like the coil and discrete_input.

Nippey replied:
Updated to behave the same way as Coil/Discrete

Nippey committed:
Update binary_sensor.py

As for Coil and Discrete, Holding should just evaluate the lowest bit of the register

Nippey commented:
> Test are missing, please amend the relevant test_file with tests that control the new functionality.

I just found and checked the test file.
It left me rather confused.
Sorry, I will need some time to understand the structure 🙈

Nippey committed:
Added test for new register type of binary sensor

Nippey commented:
@janiversen:
I think I got it.
But I am not sure, if this is already all, the test should be doing.

janiversen reviewed (changes requested):
All good apart from the CI error in isort, please solve.

Nippey committed:
Comply to pre-commit hook 'isort'

Nippey commented:
Sorry, tried to rename the branch to something more meaningful. 
But this closed the pull-request.
Keeping "patch-1".

Solved 'isort' error.

janiversen reviewed (approved):
LGTM.
